src/backend/env/memory.rs

```diff
@@ -82,50 +84,68 @@ impl Memory {
         self.api_ref.as_ref().borrow().allocator.health(unit)
     }
 
-    fn pack(&mut self) {
+    fn persist_allocator(&mut self) {
         self.api = (*self.api_ref.as_ref().borrow()).clone();
     }
 
-    fn unpack(&mut self) {
+    /// Initializes the memory allocator.
+    pub fn init(&mut self) {
         self.api_ref = Rc::new(RefCell::new(self.api.clone()));
         self.posts.init(Rc::clone(&self.api_ref));
         self.features.init(Rc::clone(&self.api_ref));
+        self.ledger.init(Rc::clone(&self.api_ref));
     }
 
     #[allow(clippy::type_complexity)]
     #[cfg(test)]
-    pub fn set_test_api(
-        &mut self,
-        mem_grow: Box<dyn FnMut(u64) -> Result<u64, String>>,
-        mem_end: Box<dyn Fn() -> u64>,
-        write_bytes: Box<dyn Fn(u64, &[u8])>,
-        read_bytes: Box<dyn Fn(u64, &mut [u8])>,
-    ) {
+    pub fn init_test_api(&mut self) {
+        static mut MEM_END: u64 = 16;
+        static mut MEMORY: Option<Vec<u8>> = None;
+        unsafe {
+            let size = 1024 * 512;
+            MEMORY = Some(Vec::with_capacity(size));
+            for _ in 0..size {
+                MEMORY.as_mut().unwrap().push(0);
+            }
+        };
+        let mem_grow = |n| unsafe {
+            MEM_END += n;
+            Ok(0)
+        };
+        fn mem_end() -> u64 {
+            unsafe { MEM_END }
+        }
+        let writer = |offset, buf: &[u8]| {
+            buf.iter().enumerate().for_each(|(i, byte)| unsafe {
+                MEMORY.as_mut().unwrap()[offset as usize + i] = *byte
+            });
+        };
+        let reader = |offset, buf: &mut [u8]| {
+            for (i, b) in buf.iter_mut().enumerate() {
+                *b = unsafe { MEMORY.as_ref().unwrap()[offset as usize + i] }
+            }
+        };
         let allocator = Allocator {
             block_size_bytes: 1,
             segments: Default::default(),
-            mem_grow: Some(mem_grow),
-            mem_size: Some(mem_end),
+            mem_grow: Some(Box::new(mem_grow)),
+            mem_size: Some(Box::new(mem_end)),
             boundary: 16,
         };
         let test_api = Api {
             allocator,
-            write_bytes: Some(write_bytes),
-            read_bytes: Some(read_bytes),
+            write_bytes: Some(Box::new(writer)),
+            read_bytes: Some(Box::new(reader)),
         };
         self.api_ref = Rc::new(RefCell::new(test_api));
         self.posts.init(Rc::clone(&self.api_ref));
         self.features.init(Rc::clone(&self.api_ref));
-    }
-
-    #[cfg(test)]
-    pub fn unpack_for_testing(&mut self) {
-        self.unpack();
+        self.ledger.init(Rc::clone(&self.api_ref));
```

Comment:
Couldn't we just `self.api = test_api` and then call `self.init()` here?

Reply:
It turns out we could not because the initialization of the allocator was happening during cloning, which is a very convoluted logic, so I refactored this part and now we can. Thanks Christian!